network/src/consensus.rs

```diff
@@ -152,24 +152,30 @@ impl<P, E> Network for ConsensusNetwork<P,E> where
 
 		// spin up a task in the background that processes all incoming statements
 		// TODO: propagate statements on a timer?
-		let inner_stream = self.network.consensus_gossip().write().messages_for(attestation_topic);
-		let process_task = self.network
-			.with_spec(|spec, ctx| {
-				spec.new_consensus(ctx, parent_hash, CurrentConsensus {
-					knowledge,
-					local_session_key,
-				});
-
-				MessageProcessTask {
-					inner_stream,
-					parent_hash,
-					table_router: table_router.clone(),
-					exit,
-				}
-			})
-			.then(|_| Ok(()));
-
-		task_executor.spawn(process_task);
+		let (tx, rx) = ::futures::sync::oneshot::channel();
+		self.network.with_gossip(move |gossip, _| {
+			let inner_rx = gossip.messages_for(attestation_topic);
+			let _ = tx.send(inner_rx);
+		});
+
+		let table_router_clone = table_router.clone();
+		let executor = task_executor.clone();
+		rx.map(move |inner_stream| {
+			self.network
+				.with_spec(move |spec, ctx| {
+					spec.new_consensus(ctx, parent_hash, CurrentConsensus {
+						knowledge,
+						local_session_key,
+					});
+					let process_task = MessageProcessTask {
+						inner_stream,
+						parent_hash,
+						table_router: table_router_clone,
+						exit,
+					};
+					executor.spawn(process_task);
+			});
+		}).wait().ok().expect("1. Network is running, 2. it should handle the work flow successfully");
```

Comment:
I would much rather have a `Stream` implementation which wraps the `rx` and doesn't panic. Waiting in asynchronous code is asking for trouble

Reply:
considering that we have a `struct NetworkDown` I don't think that `Network is running` is a valid expectation.

Reply:
Ok, I move things around a bit, relying on the fact that we're sending the closures as tasks to be executed sequentially on another thread, instead of doing the non-blocking with futures. 

Note that I also switched to using the `std::sync::mpsc` for the one-shot, since I'm not completely sure about the ordering guarantees of the futures oneshot(incidentally the std oneshot is the most efficient oneshot available)